docs/src/using-turing/compiler.md

```diff
@@ -0,0 +1,224 @@
+4. Enable the tracking of random variables using the data structures `VarName` and `VarInfo`
+5. Change `~` lines with a variable in `P` on the LHS to a call to an `assume`-block
+6. Change `~` lines with a variable in `D` on the LHS to a call to an `observe`-block
+7. Enable type stable automatic differentiation of the model using type parameters
+
+Let's take the following model as an example:
+```julia
+@model gauss(x = missing, y = 1.0, ::Type{TV} = Vector{Float64}) where {TV <: AbstractVector} = begin
+    if ismissing(x)
+        x = TV(undef, 3)
+    end
+    p = TV(undef, 2)
+    p[1] ~ InverseGamma(2, 3)
+    p[2] ~ Normal(0, 1.0)
+    @. x[1:2] ~ Normal(p[2], sqrt(p[1]))
+    x[3] ~ Normal()
+    y ~ Normal(p[2], sqrt(p[1]))
+end
+```
+A `model::Model` can be defined using `gauss(rand(3), 1.0)` or `gauss(x = rand(3), y = 1.0)`. While constructing the model, if an argument is not passed in, it will be assigned to its default value. If there is no default value given, an error will be thrown. If an argument has a default value `missing`, when not passed in, it will be treated as a random variable. For variables which require an intialization because we need to loop or broadcast over its elements, such as `x` above, the following needs to be done:
+```julia
+if ismissing(x)
+    x = ...
+end
+```
+If `x` will be sampled as a whole from a multivariate distribution, e.g. `x ~ MvNormal(...)`, there is no need to initialize it in an `if`-block.
+
+The `@model` macro is defined as:
+```julia
+macro model(input_expr)
+    build_model_info(input_expr) |> replace_tilde! |> replace_vi! |> 
+        replace_logpdf! |> replace_sampler! |> build_output
+end
+```
+
+## `build_model_info`
+
+The first stop that the model definition takes is `build_model_info`. This function extracts some information from the model definition such as:
+- `name`: the model name.
+- `main_body`: the model body excluding the header and `end`.
+- `arg_syms`: the argument symbols, e.g. `[:x, :y, :TV]` above.
+- `args`: a modified version of the arguments changing `::Type{TV}=Vector{Float64}` and `where {TV <: AbstractVector}` to `TV::Type{<:AbstractVector}=Vector{Float64}`. This is `[:(x = missing) :(y = 1.0), :(TV::Type{<:AbstractVector}=Vector{Float64})]` in the example above.
+- `args_nt`: an expression constructing a `NamedTuple` of the input arguments, e.g. :((x = x, y = y, T = T)) in the example above.
+and returns it as a dictionary called `model_info`.
+
+## `replace_tilde!`
+
+After some model information have been extracted, `replace_tilde!` replaces the `L ~ R` lines in the model with the output of `tilde(L, R, model_info)` where `L` and `R` are either expressions or symbols. `L` can also be a constant literal. The `replace_tilde!` function also replaces expressions of the form `@. L ~ R` with the output of `dot_tilde(L, R, model_info)`.
```

Comment:
Any reason why `model_info` is passed to `tilde`?

Reply:
This is `Core.tilde` which is different from `Inference.tilde`. `Core.tilde` returns the expression that will be run instead of the `~` line. To write out this expression, we need the variable names generated for `model`, `ctx`, `vi`, and `sampler`. These names are `gensym`ed in `build_model_info` and saved in `model_info` for reuse. One part of the expression generated by `Core.tilde` is a call to `Inference.tilde`. `Core.tilde` is defined in the `compiler.jl` file, while `Inference.tilde` is defined in the `Inference.jl` file.

Reply:
Thanks for the clarification. Maybe add this distinction to the docs, e.g. in the form of a comment/footnote?

Reply:
Sure